Add category fields to the network admin new site form

hvianna opened the pull request:
* Replaced wp_get_sites() which is deprecated since 4.6 and calls
get_sites() instead. The new function returns a maximum of 100 sites by
default, so the parameter 'number' => null was added to retrieve all
sites in networks larger than 100 sites;

* Added action for 'network_site_new_form' so the category fields are
also displayed in the network admin new site form;

* The 'add_signup_meta' filter is not fired up when the new site is
created by the network admin, so I had to move the code which gets the
category form fields to the wpmu_new_blog() function. Not so fancy, but
it work for both cases, so I removed (commented out) the action/function
for 'add_signup_meta'.

Commit:
Merge branch 'develop'

git-svn-id: https://plugins.svn.wordpress.org/multisite-directory/trunk@1415224 b8457f37-d9ea-0310-8a92-e5e31aec5664

Commit:
Merge branch 'develop'

git-svn-id: https://plugins.svn.wordpress.org/multisite-directory/trunk@1474456 b8457f37-d9ea-0310-8a92-e5e31aec5664

hvianna committed:
Fix for deprecated wp_get_sites(); Add category fields to the network admin new site form.

* Replaced wp_get_sites() which is deprecated since 4.6 and calls
get_sites() instead. The new function returns a maximum of 100 sites by
default, so the parameter 'number' => null was added to retrieve all
sites in networks larger than 100 sites;

* Added action for 'network_site_new_form' so the category fields are
also displayed in the network admin new site form;

* The 'add_signup_meta' filter is not fired up when the new site is
created by the network admin, so I had to move the code which gets the
category form fields to the wpmu_new_blog() function. Not so fancy, but
it work for both cases, so I removed (commented out) the action/function
for 'add_signup_meta'.

hvianna committed:
i18n + redirect site posts to the actual sites

+ Added action/function to load the plugin's textdomain
+ Added pt_BR translation
+ Renamed previous translation files to match the textdomain
+ Added action/function to redirect single site posts to the actual
sites in the network
+ Renamed the admin menu item from 'Sites' to 'Directory', so it won't
get confused with the network admin menu item

meitar commented:
Thanks for your contribution! :) I think this would be a nice addition to the project, and at first glance everything looks really clean and beautiful. I do have some concerns about the implementation you've proposed here.

* Is there some reason why you've chosen to comment out old methods instead of deleting them? It's generally much better to simply delete code rather than to comment it out. Commented code quickly becomes confusing and is a style that, to me, generally [smells bad](https://en.wikipedia.org/wiki/Code_smell).

I'm gonna go pull these changes down myself and give them a try on my test network. Very much appreciate your work on this so far and there's definitely a lot in here worth merging. 👍

fabacab commented on multisite-directory.php:
```diff
@@ -179,19 +183,20 @@ public static function deactivate ($network_wide) {
      */
     private function initializeDirectory () {
-        $sites = wp_get_sites(array(
+        $sites = get_sites(array(
```
This is the right move, but unfortunately it means that you also have to change the `Requires` line in `readme.txt` to 4.6, since [`get_sites()` does not exist in earlier versions](https://developer.wordpress.org/reference/functions/get_sites/#changelog). I'm totally comfortable requiring WordPress 4.6 now, so there's no need to provide a backwards compatibility check. But we will need to update the `Requires` line or I will have Very Unhappy Users in [the support forum](https://wordpress.org/support/plugin/multisite-directory). Thanks!

hvianna replied:
Updated!

fabacab commented on multisite-directory.php:
```diff
@@ -109,7 +121,10 @@ public static function register_scripts () {
      */
     public static function wpmu_new_blog ($blog_id) {
-        $signup_cats = get_blog_option($blog_id, 'multisite-directory-signup-categories');
+        if (isset($_POST['tax_input']) && !empty($_POST['tax_input'][Multisite_Directory_Taxonomy::name])) {
```
It's cool that this works in both cases where the input is being sent by the network's signup form as well as in the WP Network Admin screens, but it would be nice if this were more explicitly documented. Would you consider adding a docblock comment along the lines of `@uses $_POST['tax_input'] to retrieve the category fields during new site creation on either the signup front-end or the Network Admin screen`. This way, we know what it's there for. It has the added benefit of informing us that we're dealing with a global here, too.

hvianna replied:
Added as you suggested, please feel free to adjust the text as you see fit.

fabacab commented on multisite-directory.php:
```diff
@@ -220,12 +236,14 @@ public static function signup_blogform () {
      *
      * @param array $meta
      */
+/*
```
Rather than commenting out old code, just delete it. No need to keep old code in comments. :)

hvianna replied:
Done!

fabacab commented on multisite-directory.php:
```diff
@@ -49,7 +50,9 @@ public static function register () {
         add_action('network_admin_menu', array('WP_Multisite_Directory_Admin', 'network_admin_menu'));
         add_action('signup_blogform', array(__CLASS__, 'signup_blogform'));
-        add_action('add_signup_meta', array(__CLASS__, 'add_signup_meta'));
+        add_action('network_site_new_form', array(__CLASS__, 'signup_blogform'));
```
Since the `signup_blogform()` method is now being used in more than one place (i.e., it's not only being used on the new site sign-up form "on the front-end," you should update [its documentation (its docblock comment)](https://github.com/hvianna/multisite-directory/blob/7c39e53ad8fde6f075146795d3e22ecbe4e12d39/multisite-directory.php#L216) to reflect this fact.

An even better improvement would be to use a different method for the `network_site_new_form` action altogether, extract the parts of the existing `signup_blogform()` method that are generic into a helper function (although right now that's just one line, so that's possibly overkill), and use this new helper method in the new method. The purpose of the new method would be to output HTML that matches its context.

Currently, `signup_blogform()` creates HTML that's geared for the New Blog signup form on the front-end, but on the Network Admin screen, it sticks out like a sore thumb:

<img width="761" alt="screen shot 2016-12-14 at dec 14 11 23 04 am" src="https://cloud.githubusercontent.com/assets/18677/21190506/00d6c2d0-c1f0-11e6-80f3-ceeed4cc5d46.png">

There are ready-made HTML and CSS conventions that the WordPress admin areas use for this, just inspect the HTML of the fields above and match that output in the new method. :) (If this seems too complex, no stress, we can always do it after merging this PR. The only thing that really *must* be changed here before merging is updating the docblock comment for the given method so the documentation doesn't become factually incorrect.)

hvianna replied:
I added a check using the `$pagenow` global to detect whether the user is on the network admin screen and output the additional HTML. Not sure if you're gonna be happy with that approach though, as your suggestion would surely be more elegant. I also added a docblock `@uses` to document the usage of the global variable - please feel free to adjust the language if it doesn't feel right, I'm not actually fluent in English.

I think we could get rid of the list bullets (they were already visible on the front-end sign-up form). Do you think injecting an inline style in the checklist `$html` would be too smelly? :)

hvianna replied:
About the bullets, I just realized that `wp_terms_checklist` doesn't include a `<ul>` in the code it returns, so we could just add a `<ul style="list-style:none;">` to the HTML  output.

fabacab commented on multisite-directory.php:
```diff
@@ -246,6 +264,16 @@ public static function dashboard_glance_items ($items) {
+     * Redirects single site post to the actual site
+     */
+    public static function template_redirect () {
+        if (is_singular(Multisite_Directory_Entry::name)) {
```
I like the idea of more quickly navigating from directory entry to the site itself, but this redirect needs to be optional. Part of the point of having the directory entries be posts, themselves, is for the Network Admin to write long-form descriptions of the sites in the network. With this, you are *unconditionally* redirecting all page views of the entry itself to the site that the entry refers to. That isn't the intended purpose of the plugin.

The best way to do this is to store a `post_meta` value in the entry itself that is then checked on this line to see if a redirect should or should not happen, and an option in the main blog's `options` table to check whether this should be active by default or not. (It should default to the same behavior before this patch, meaning it should be off by default.) That way Network Admins can turn on or off the redirect for each individual site in their Directory and can configure the default for new sites.

This is a bit more work, admittedly, and so you may want to remove these changes from #19 and submit a new pull request with these changes. In the mean time, you can [open an issue](https://github.com/meitar/multisite-directory/issues/new) requesting this specific feature so we have a place to hash it out, if needed.

hvianna replied:
You're right, I completely overlooked that! I'm removing this for now and will open the issue, but your solution sounds just perfect!

fabacab commented on multisite-directory.php:
```diff
@@ -49,7 +50,9 @@ public static function register () {
         add_action('signup_blogform', array(__CLASS__, 'signup_blogform'));
-        add_action('add_signup_meta', array(__CLASS__, 'add_signup_meta'));
+        add_action('network_site_new_form', array(__CLASS__, 'signup_blogform'));
+//        add_action('add_signup_meta', array(__CLASS__, 'add_signup_meta'));
```
As before, this line shouldn't be commented out, it should just be deleted. :)

hvianna replied:
Done!

fabacab reviewed (commented):
One more point about this pull request: the base branch of this pull request should be `develop`, not `master`. Since this is a WordPress plugin, I use the `master` branch as a `git svn` remote, since the WP Plugin Directory *only* supports Subversion, and Subversion doesn't play nice when I try to make it handle Git's merge commits. :\

hvianna commented:
Thank you for the super positive response! I'm really glad this will actually be useful to the project! :D 

I completely agree with you about the commented out parts, but I was a bit hesitant to simply delete your original code. As a bonus, TIL about code smell! :P

I'll go thru your recommendations and will try to fix everything tomorrow morning.

hvianna committed:
Fixed issues and updated documentation

Addressed issues discussed and removed the automatic redirection of
directory entries to the actual sites.

fabacab commented on multisite-directory.php:
```diff
@@ -213,37 +212,37 @@ private function initializeDirectory () {
-        if (isset($_POST['tax_input']) && !empty($_POST['tax_input'][Multisite_Directory_Taxonomy::name])) {
-            $meta['multisite-directory-signup-categories'] = $_POST['tax_input'][Multisite_Directory_Taxonomy::name];
+        print '<div id="multisite-directory-signup-categories">';
+        if ($pagenow == 'site-new.php') {
```
Just a minor note, if you're not familiar with this convention, consider reading about "[Yoda conditions](https://en.wikipedia.org/wiki/Yoda_conditions)." The short version is that:

```php
'site-new.php' == $pagenow
```

is "better" than the other way around (the way you have it):

```php
$pagenow == 'site-new.php'
```

The reason it's "better" with the variable to the *right* of the conditional operator is because you protect the code from an accidental deletion of one of the `=` characters, which would change the meaning of the statement from a condition to an assignment (`$pagenow = 'site-new.php'`). With the variable to the right of the conditional, this same mistake would cause a fatal error rather than re-assigning to the `$pagenow` variable, because you cannot assign a variable to a string literal: `'site-new.php' = $pagenow // fatal error`.

This kind of defensive programming is especially important when dealing with global variables. [Yoda conditions in PHP are also a WordPress coding standard](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#yoda-conditions).

Anyway, I don't think you need to change this per se, I've already marked it on my local branch for modification. I'm just giving you a heads up about it in case you don't already know about Yoda  conditions so you can start using them in the future and improve your own coding practice.

hvianna replied:
Oh yes, I'm familiar with the technique but didn't know people called it "Yoda conditions". Now somehow it seems more appealing :D I just don't like the way it looks in the code, but I can understand the benefits. I also wasn't aware that it is a WP coding standard. I'll include this change in the final commit.

meitar commented:
This looks almost ready to merge now! Thanks again @hvianna! Last things to do before merging is to clean up your commit history. Since you've already got a few non-atomic commits, the easiest thing to do is just squash all of these together into one commit. Follow these instructions if you don't already know what that means.

1. Make a topic branch so your current work (which seems to be on `master`) can be referenced somewhere other than `master`.
1. Since you should be committing on top of `develop`, not `master`, checkout `develop`
1. Now you can squash all your work into one single commit.
1. And finally you should commit this.

The [`git(1)`](https://www.kernel.org/pub/software/scm/git/docs/) commands for this procedure might look like the following:

```sh
git branch my-pull-request master      # create a new branch called "my-pull-request" at the same place as "master" currently is
git checkout -b develop origin/develop # create a local develop branch that tracks the "develop" branch in your remote called "origin"
git merge --squash my-pull-request     # squash all commits present in "my-pull-request" into one commit
git commit -v                          # actually commit the single, squashed commit object prepared by the previous command
```

At this point your local `develop` branch will contain the contents you actually want to submit as your pull request, so you should `git push` and update this pull request's head branch to reference that. GitHub will then show only 1 commit object to merge and it will be a clean merge. I can then merge your pull request into my `develop` branch and from there merge it into my local `master` branch to update WordPress's Subversion side of things in a way that won't freak it out because we've been using Git all along and not Subversion like it expects.

Thanks again!

hvianna committed:
Removed bullets from category checklist

Added HTML/CSS to remove the list bullets from the category checklist.

hvianna committed:
"You must unlearn what you have learned."

Changed $pagenow check to a "Yoda condition".

hvianna commented:
I could not find how to change my branch to 'develop', so I opened a new pull request.

meitar commented:
ba97555b9aa8e6e36ef3628f165796ff41f57a60 has the best commit message ever. :)
